### PR TITLE
Store `Ident` in `DefPathData` instead of `Symbol`

### PR DESCRIPTION
**Pull request by Aaron1011**
This allows us to recover span information when emitting cross crate
errors. A number of test cases benefit from this change, since we were
previously not emitting messages due to invalid spans.

There are four main parts to this commit:

1. Adding `Ident` to `DefPathData`, and updating the affected code. This
mostly consists of mechanical changes.

2. Updating how we determine the disambiguator for a `DefPath`. Since
`DefPathData` now stores a `Span` (inside the `Ident`), we need to
make sure we ignore this span we determining if a path needs to be
disambiguated. This ensure that two paths with the same symbols but
different spans are considered equal (this can occur when a macro is
repeatedly expanded to a definition).

3. Ensuring that we are able to decode `Spans` when decoding the
`DefPathTable`. Since the `DefPathTable` is stored in `CrateMetadata`, we
must decode it before we have a `CrateMetadata` available. Since
decoding a `Span` requires access to several fields from
`CrateMetadata`, this commit adds a new struct `InitialCrateMetadata`,
which implements `Metadata` and stores all of the necessary fields.

4. The specialized metadata encoder/decoder impls for `Ident` are
removed. This causes us to fall back to the default encoder/decoder
implementations for `Ident`, which simply serializes and deserializes
the fields of `Ident`. This is strictly an improvement - we still don't
have any hygiene information, but we now have a non-dummy Span.

This should hopefully allow us to test PR #68941, since we will now use
cross-crate spans in more places.

### CODE REVIEW

File: src/test/ui/pattern/usefulness/match-arm-statics-2.stderr [hunk 1]
```diff
@@ -11,6 +11,14 @@ error[E0004]: non-exhaustive patterns: `Some(Some(West))` not covered
...
 LL |     match Some(Some(North)) {
    |           ^^^^^^^^^^^^^^^^^ pattern `Some(Some(West))` not covered
+   | 
+  ::: $SRC_DIR/libcore/option.rs:LL:COL
+   |
+LL |     Some(#[stable(feature = "rust1", since = "1.0.0")] T),
+   |     ----
+   |     |
+   |     not covered
+   |     not covered
```
Comment (Aaron1011) [inline src/test/ui/pattern/usefulness/match-arm-statics-2.stderr]: This is a little confusing, but it also occurs for enums defined in the same crates, so it isn't an issue with this PR.

**Comment by bjorn3** [comment]
Won't this regress incremental compilation, as changing the span of a definition will change it's `DefPath`, thus producing a definition that is different from the point of view of the incremental system.

**Comment by Aaron1011** [comment]
@bjorn3: The `DefPathHash` still only use symbol names, not spans (hence the need to explicitly ignore spans when determining the didambiguator, as the paths could otherwise be identical).

**Comment by Aaron1011** [comment]
@bjorn3: Note that we already associate a `Span` with each definition - this just adds a `Span` that specifically points to the identifier. I don't think this will make incremental compilation any more dependent on `Span`s than it already is - but it might be good to get a perf run to check.

**Comment by petrochenkov** [comment]
@Aaron1011 
>Note that we already associate a Span with each definition - this just adds a Span that specifically points to the identifier.

Can this new span be kept somewhere near that old span instead of what this PR does?

The purpose of `DefPathData` is to serve as a key in some table, and expanding it with some data that has to be specifically excluded before that key can be compared or hashed doesn't look like a good idea.

File: src/librustc/hir/map/definitions.rs [hunk 2]
```diff
@@ -537,22 +560,22 @@ impl DefPathData {
         }
     }
 
-    pub fn as_symbol(&self) -> Symbol {
```
Comment (petrochenkov) [inline src/librustc/hir/map/definitions.rs]: If the direction with keeping the span in `DefPathData` is pursued, this method can be kept for compatibility and used in places where full identifier is not needed. That will significantly reduce the churn.

**Comment by Aaron1011** [comment]
> Can this new span be kept somewhere near that old span instead of what this PR does?

The span is logically associated with the `Symbol` stored in the `DefPathData` variant - we're essentially just holding on to an `Ident` that we were previously throwing away when constructing a `DefPathData`.

Storing the span somewhere else would either require us to duplicate the variants of `DefPathData`, or create an invalid span for variants that don't have a `Symbol` (e.g. `ClosureExpr`). Both seem kind of hacky.

> The purpose of DefPathData is to serve as a key in some table, and expanding it with some data that has to be specifically excluded before that key can be compared or hashed doesn't look like a good idea.

I agree that this isn't great. I'm going to see how much fallout there is from removing the `Hash`, `Eq`, and `PartialEq` impls, and adjusting the `next_disambiguator` map.

File: src/librustc_resolve/lib.rs [hunk 3]
```diff
@@ -835,7 +834,7 @@ pub struct Resolver<'a> {
 
     /// Names of fields of an item `DefId` accessible with dot syntax.
     /// Used for hints during error reporting.
-    field_names: FxHashMap<DefId, Vec<Spanned<Name>>>,
+    field_names: FxHashMap<DefId, Vec<Ident>>,
```
Comment (petrochenkov) [inline src/librustc_resolve/lib.rs]: This is not an identifier, it's a `Name` combined with an unrelated `Span`.
If we want to avoid `Spanned`, then this can be e.g. `(Name, Span)`.
The change is unrelated to this PR though.

Comment (Aaron1011) [reply src/librustc_resolve/lib.rs]: The `Span` is very much related - it's the exact span where the `Symbol` occurred (if it's a named field), or the span of the type position (if it's an unnamed field).

However, I see your point about this lacking the additional meaning of `Ident` (the second case, strictly speaking, isn't an identifier). I'll change this back.

**Comment by eddyb** [comment]
cc @nikomatsakis @michaelwoerister @Zoxc

**Comment by Aaron1011** [comment]
@petrochenkov: I removed the `Eq`, `PartialEq`, and `Hash` impls from `DefPathData`. This results in a small amount of boilerplate within `definitions.rs` (I added a `PlainDefPathData` enum that's essentially a copy of `DefPathData`). However, it allows the rest of the codebase to deal with `Idents` (which is the proper type to be storing), instead of needing to combine a `Symbol` with a `Span` retrieved from elsewhere.

There is no longer any possibility of accidentally comparing two `DefPathData`s based on their `Span`s.

**Comment by Zoxc** [comment]
This is problematic for incremental compilation. You can't include `Span` here without also hashing it (since hashing is used to approximate equality).

**Comment by Aaron1011** [comment]
@Zoxc: What would be the consequences of not hashing the `Span`? The `Span` should only be used when printing error messages - how does that interact with incremental compilation?

**Comment by Aaron1011** [comment]
Note that storing these `Ident`s in *some* form is necessary for cross-crate hygiene to work properly. When we encode something like:

```rust
#![feature(decl_macro)]
macro x() { struct X; }

x!();
x!();
```

we need to encode the full `Ident`, not just the `Symbol`, for both definitions of `X` (so that we encode the different `SyntaxContext`s).

This means that we actually do care about the `Ident`'s `Span` (at least the `SyntaxContext`) in places other than diagnostics.

Does storing the `Span` in a side table vs using an `Ident` make a difference in terms of incremental compilation?

**Comment by bjorn3** [comment]
Storing a Span in a side table automatically does the correct thing: Only add a dependency when it is actually used. Storing an Ident in the DefPath requires hashing to ignore it to avoid a dependency everywhere and that also means that there is no dependency when there should be one. If the macro `x` were defined in a different module and contained a function and you moved it, the cgu containing the function declared by `x!()` should be recompiled to make the debuginfo match.

**Comment by eddyb** [comment]
Funnily enough, there's already a `def_id_to_span` (or I guess `def_index_to_span` pre-#66131) in `Definitions`, but I suppose that has the entire definition `Span`, not just the name `Span`.

**Comment by Zoxc** [comment]
@Aaron1011 Does your example result in two distinct `DefId`s today? Or is that something that needs fixing.

**Comment by Aaron1011** [comment]
@Zoxc: My example does create two distinct `DefIds`. However, when we serializes the crate metadata, both structs end up with `SyntaxContext::root()` in their identifier span. This causes a resolution error when we try to import these items from a different crate, since both items end up with identical `Ident`s.

**Comment by Aaron1011** [comment]
Closing in favor of https://github.com/rust-lang/rust/pull/70077, which uses the side table approach.
